File: crates/verify/ess-conformance/src/go/live_trace.go
// LiveObservationTarget supplies complete facts on one subscribed monotonic lifetime.
// EndScenario owns release, including partially opened subscriptions.
type LiveObservationTarget interface {
	OpenObservations(events []string) (string, error)
	ObserveOccurrences(request LiveObservationRequest) (ObservationBatch, error)
}

type LiveObservationRequest struct {
	Lifetime  string
	After     uint64
	ThroughMS uint64
}

type LiveClaim struct {
	Event   string           `json:"event"`
	Matches map[string]Value `json:"matches"`
	Shape   map[string]Held  `json:"shape"`
}

type LiveCheck struct {
	Instance string `json:"instance"`
	Source string `json:"source"`
	Event string `json:"event"`
	Path string `json:"path"`
	RequirePresent bool `json:"require_present"`
	Min int64 `json:"min"`
	Max int64 `json:"max"`
	Plus int64 `json:"plus"`
	Kind       string           `json:"kind"`
	Events     []string         `json:"events"`
	Anchor     string           `json:"anchor"`
	After      *string          `json:"after"`
	Before     string           `json:"before"`
	Claim      *LiveClaim       `json:"claim"`
	DurationMS uint64           `json:"duration_ms"`
	Required   map[string]Value `json:"required"`
	Shape      map[string]Held  `json:"shape"`
}

type liveTraceState struct {
	ledger  *ObservationLedger
	anchors map[string]uint64
	events  map[string]bool
	quietAnchor string
	quietFloor *uint64
}

func admitLiveTrace(value any) error {
	f, ok := value.(map[string]any)
	if !ok {
		return fmt.Errorf("live trace must be an object")
	}
	kind, err := text(f["kind"])
	if err != nil {
		return err
	}
	required := "kind"
	switch kind {
	case "capture":
		required += " anchor instance event path shape require_present"
	case "offset":
		required += " instance source plus"
	case "integer_bounds":
		required += " anchor event path shape min max"
	case "quiet":
		required += " anchor duration_ms claim"
	case "open":
		required += " events"
	case "await":
		required += " anchor after claim"
	case "absent":
		required += " anchor duration_ms claim"
	case "stable":
		required += " anchor duration_ms claim required shape"
	case "ordered":
		required += " before after"
	default:
		return fmt.Errorf("unknown live trace kind")
	}
	if _, err := closed(value, required, ""); err != nil {
		return err
	}
	if raw, exists := f["shape"]; exists {
		if err := admitShape(raw); err != nil { return err }
	}
	if raw, exists := f["event"]; exists {
		if err := name(raw, false); err != nil { return err }
	}
	if raw, exists := f["claim"]; exists {
		claim, err := closed(raw, "event matches shape", "")
		if err != nil {
			return err
		}
		if err := name(claim["event"], false); err != nil {
			return err
		}
		if err := admitShape(claim["shape"]); err != nil {
			return err
		}
		if _, ok := claim["matches"].(map[string]any); !ok {
			return fmt.Errorf("live matches must be an object")
		}
		if err := admitValues(claim["matches"], 10, false); err != nil {
			return err
		}
	}
	if raw, exists := f["required"]; exists {
		required, ok := raw.(map[string]any)
		if !ok {
			return fmt.Errorf("live required must be an object")
		}
		if err := admitValues(required, 10, false); err != nil {
			return err
		}
		if err := admitShape(f["shape"]); err != nil {
			return err
		}
	}
	for _, field := range []string{"anchor", "before", "after", "instance", "source"} {
		if v, exists := f[field]; exists && !(kind == "await" && field == "after" && v == nil) {
			if err := name(v, true); err != nil {
				return err
			}
		}
	}
	encoded, err := json.Marshal(value)
	if err != nil {
		return err
	}
	var check LiveCheck
	decoder := json.NewDecoder(strings.NewReader(string(encoded)))
	decoder.UseNumber()
	if err := decoder.Decode(&check); err != nil {
		return err
	}
	if kind == "capture" || kind == "integer_bounds" {
		leaf, exists := check.Shape[check.Path]
		if !exists || len(check.Shape) != 1 || check.Path == "" {
			return fmt.Errorf("event capture needs exactly one declared payload path")
		}
		for _, part := range strings.Split(check.Path, ".") {
			if part == "" { return fmt.Errorf("empty observed payload path segment") }
		}
		if kind == "capture" {
			if _, ok := f["require_present"].(bool); !ok || (leaf.Optional && !check.RequirePresent) {
				return fmt.Errorf("optional event capture requires explicit presence")
			}
		} else if !liveIntegerLeaf(leaf) || check.Min > check.Max {
			return fmt.Errorf("integer bounds require a declared Integer and ordered range")
		}
	}
	if kind == "open" {
		if len(check.Events) == 0 {
			return fmt.Errorf("empty observation subscription")
		}
		seen := map[string]bool{}
		for _, event := range check.Events {
			if !qualifiedName.MatchString(event) || seen[event] {
				return fmt.Errorf("invalid or duplicate subscription event")
			}
			seen[event] = true
		}
	}
	if (kind == "absent" || kind == "stable" || kind == "quiet") && check.DurationMS == 0 {
		return fmt.Errorf("live window must be positive")
	}
	return nil
}

func admitLiveSequence(steps []any, major int) error {
	events, captures := map[string]bool{}, map[string]bool{}
	anchors := map[string]string{}
	integers := map[string]bool{}
	for index, raw := range steps {
		data, _ := json.Marshal(raw)
		// Parent admission must not narrow unrelated legacy u64 metadata to
		// execution's machine-sized int fields. Decode only this vocabulary.
		var step struct {
			Step     string     `json:"step"`
			Instance string     `json:"instance"`
			Trace    *LiveCheck `json:"trace"`
			Field string `json:"field"`
			Shape map[string]Held `json:"shape"`
		}
		decoder := json.NewDecoder(strings.NewReader(string(data)))
		decoder.UseNumber()
		if err := decoder.Decode(&step); err != nil {
			return err
		}
		switch step.Step {
		case "capture_response", "capture_instance", "establish_entity":
			captures[step.Instance] = true
			if step.Step == "capture_response" && liveIntegerLeaf(step.Shape[step.Field]) {
				integers[step.Instance] = true
			}
		case "check_live":
			if major < 10 {
				return fmt.Errorf("live trace requires suite/10 or /11")
			}
			c := step.Trace
			if major < 12 && (c.Kind == "capture" || c.Kind == "offset" ||
				c.Kind == "integer_bounds" || c.Kind == "quiet") {
				return fmt.Errorf("captured observations require suite/12 or /13")
			}
			if c.Kind == "open" {
				if index != 0 {
					return fmt.Errorf("observation subscription must be first")
				}
				for _, event := range c.Events {
					events[event] = true
				}
				continue
			}
			if len(events) == 0 {
				return fmt.Errorf("live assertion before subscription")
			}
			if c.Event != "" && !events[c.Event] {
				return fmt.Errorf("observed event was not subscribed")
			}
			if c.Claim != nil {
				if !events[c.Claim.Event] {
					return fmt.Errorf("assertion event was not subscribed")
				}
				if err := admitLiveClaimValues(c.Claim.Matches, c.Claim.Shape, captures); err != nil {
					return err
				}
			}
			switch c.Kind {
			case "capture":
				if anchors[c.Anchor] != c.Event || captures[c.Instance] {
					return fmt.Errorf("capture needs its actual event and a unique value name")
				}
				captures[c.Instance] = true
				integers[c.Instance] = liveIntegerLeaf(c.Shape[c.Path])
			case "offset":
				if !integers[c.Source] || captures[c.Instance] {
					return fmt.Errorf("offset needs a preceding typed integer and unique value name")
				}
				captures[c.Instance], integers[c.Instance] = true, true
			case "integer_bounds":
				if anchors[c.Anchor] != c.Event {
					return fmt.Errorf("integer bound event differs from its occurrence")
				}
			case "quiet":
				if anchors[c.Anchor] != "" { return fmt.Errorf("duplicate occurrence anchor") }
				anchors[c.Anchor] = c.Claim.Event
			case "await":
				if c.After != nil && anchors[*c.After] == "" {
					return fmt.Errorf("unknown occurrence anchor")
				}
				if anchors[c.Anchor] != "" {
					return fmt.Errorf("duplicate occurrence anchor")
				}
				anchors[c.Anchor] = c.Claim.Event
			case "absent", "stable":
				if anchors[c.Anchor] == "" {
					return fmt.Errorf("unknown occurrence anchor")
				}
				if c.Kind == "stable" {
					if err := admitLiveClaimValues(c.Required, c.Shape, captures); err != nil {
						return err
					}
					for a := range c.Required {
						for b := range c.Claim.Matches {
							if a == b || strings.HasPrefix(a, b+".") || strings.HasPrefix(b, a+".") {
								return fmt.Errorf("stability claim overlaps scope filter")
							}
						}
					}
				}
			case "ordered":
				if anchors[c.Before] == "" || c.After == nil || anchors[*c.After] == "" {
					return fmt.Errorf("unknown occurrence anchor")
				}
			}
		}
	}
	return nil
}

func liveIntegerLeaf(leaf Held) bool {
	return leaf.Holds == "primitive" && leaf.Kind == "integer"
}

func liveInteger(value Node) (int64, bool) {
	number, ok := responseNumber(value)
	if !ok || !number.IsInt() || !number.Num().IsInt64() { return 0, false }
	return number.Num().Int64(), true
}

func admitLiveClaimValues(values map[string]Value, shape map[string]Held, captures map[string]bool) error {
	if len(values) == 0 {
		return fmt.Errorf("empty live payload claim")
	}
	for path, value := range values {
		leaf, present := shape[path]
		if !present {
			return fmt.Errorf("live path has no declared type")
		}
		if value.Kind == "literal" && liveLeafAdmits(leaf, value.Value, true) {
			continue
		}
		if value.Kind == "instance" && captures[value.Instance] {
			continue
		}
		return fmt.Errorf("live claim requires typed literal or preceding capture")
	}
	return nil
}

func (r *run) checkLive(index int, c *LiveCheck) bool {
	target, ok := r.target.(LiveObservationTarget)
	if !ok {
		r.skip("target cannot provide complete occurrence observations")
		return false
	}
	fail := func(err error) bool {
		if errors.Is(err, ErrUnsupported) {
			r.skip("live observation unsupported: %v", err)
			return false
		}
		return r.fail(index, "live observation: %v", err)
	}
	if c.Kind == "open" {
		lifetime, err := target.OpenObservations(append([]string(nil), c.Events...))
		if err != nil {
			return fail(err)
		}
		ledger, err := NewObservationLedger(lifetime)
		if err != nil {
			return fail(err)
		}
		r.trace = &liveTraceState{ledger: ledger, anchors: map[string]uint64{}, events: map[string]bool{}}
		for _, event := range c.Events {
			r.trace.events[event] = true
		}
		return true
	}
	state := r.trace
	var matcher OccurrenceMatcher
	if c.Claim != nil {
		values, ok := r.resolveAll(index, c.Claim.Matches)
		if !ok {
			return false
		}
		matcher = OccurrenceMatcher{Event: c.Claim.Event, Payload: values}
	}
	required, ok := r.resolveAll(index, c.Required)
	if !ok {
		return false
	}
	_, bounded := context.Background().Deadline()
	if r.ctx != nil {
		_, bounded = r.ctx.Deadline()
	}
	for attempt := 0; ; attempt++ {
		if r.ctx != nil && r.ctx.Err() != nil {
			return fail(r.ctx.Err())
		}
		var err error
		switch c.Kind {
		case "capture", "integer_bounds":
			var item *Occurrence
			item, err = state.ledger.anchor(state.anchors[c.Anchor])
			if err != nil { break }
			value, present := lookup(item.Payload, c.Path)
			valid := item.Event == c.Event && present && value != nil
			if c.Kind == "integer_bounds" {
				count, ok := liveInteger(value)
				valid = valid && ok && c.Min <= count && count <= c.Max
			} else {
				valid = valid && holdsLive(item.Payload, c.Shape) == ""
			}
			if !valid {
				err = &TemporalError{Kind: "counterexample", Counterexample: copyOccurrence(*item)}
				break
			}
			if c.Kind == "capture" {
				if _, exists := r.instances[c.Instance]; exists {
					err = fmt.Errorf("duplicate live capture")
					break
				}
				var copied map[string]Node
				copied, err = temporalFields(map[string]Node{"value": value})
				if err == nil { r.instances[c.Instance] = copied["value"] }
			}
		case "offset":
			base, ok := liveInteger(r.instances[c.Source])
			_, exists := r.instances[c.Instance]
			if !ok || exists {
				err = fmt.Errorf("offset requires a present signed integer and fresh binding")
			} else if (c.Plus > 0 && base > math.MaxInt64-c.Plus) ||
				(c.Plus < 0 && base < math.MinInt64-c.Plus) {
				err = fmt.Errorf("signed integer offset overflow")
			} else {
				r.instances[c.Instance] = json.Number(strconv.FormatInt(base+c.Plus, 10))
			}
		case "quiet":
			if state.quietAnchor != c.Anchor {
				state.quietAnchor, state.quietFloor = c.Anchor, nil
			}
			if state.quietFloor == nil {
				err = &TemporalError{Kind: "unfinished"}
				break
			}
			var found *Occurrence
			found, err = state.ledger.Quiet(matcher, *state.quietFloor, c.DurationMS)
			if err == nil {
				if violation := holdsLive(found.Payload, c.Claim.Shape); violation != "" {
					return r.fail(index, "%s", violation)
				}
				state.anchors[c.Anchor] = found.Sequence
			}
		case "await":
			after := uint64(0)
			if c.After != nil {
				after = state.anchors[*c.After]
			}
			var found *Occurrence
			found, err = state.ledger.Find(matcher, after)
			if err == nil && found == nil {
				err = &TemporalError{Kind: "unfinished"}
			}
			if err == nil {
				if violation := holdsLive(found.Payload, c.Claim.Shape); violation != "" {
					return r.fail(index, "%s", violation)
				}
				state.anchors[c.Anchor] = found.Sequence
			}
		case "absent":
			err = state.ledger.Absent(matcher, state.anchors[c.Anchor], c.DurationMS)
		case "stable":
			err = state.ledger.Stable(matcher, required, state.anchors[c.Anchor], c.DurationMS)
		case "ordered":
			err = state.ledger.Ordered(state.anchors[c.Before], state.anchors[*c.After])
		}
		if err == nil {
			return true
		}
		var detail *TemporalError
		if !errors.As(err, &detail) || detail.Kind != "unfinished" {
			return fail(err)
		}
		if !bounded && attempt >= 8 {
			return fail(err)
		}
		batch, err := target.ObserveOccurrences(LiveObservationRequest{Lifetime: state.ledger.lifetime, After: state.ledger.Cursor(), ThroughMS: detail.RequiredMS})
		if err != nil {
			return fail(err)
		}
		for _, occurrence := range batch.Occurrences {
			if !state.events[occurrence.Event] {
				return fail(state.ledger.Invalidate("unsubscribed event in observation batch"))
			}
		}
		if err := state.ledger.Accept(batch); err != nil {
			return fail(err)
		}
		if c.Kind == "quiet" && state.quietFloor == nil {
			floor := state.ledger.completeBeforeMS
			state.quietFloor = &floor
		}
	}
}

File: crates/verify/ess-conformance/src/live_trace.rs
//! Closed occurrence-based live assertions. Targets provide complete facts, never verdicts.
use crate::scenario::{EventRef, InstanceName, PayloadShape, ScenarioStep, ScenarioValue};
use crate::temporal::{self, Batch, Ledger, Matcher};
use ess_primitives::facts::Number;
use ess_primitives::node::Node;
use std::collections::{BTreeMap, BTreeSet};

/// A typed payload-path claim resolved against an original event declaration.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Claim {
    /// Original qualified event.
    pub event: EventRef,
    /// Literals and earlier actual captures at declared paths.
    pub matches: BTreeMap<String, ScenarioValue>,
    /// Original leaf types for those paths.
    pub shape: PayloadShape,
}

/// Suite/10 and /11 live checks; these versions are the unreleased live vocabulary.
#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum Check {
    /// Capture one declared field from the named actual occurrence (suite/12+).
    Capture {
        /// Previously observed occurrence.
        anchor: InstanceName,
        /// Unique value binding.
        instance: InstanceName,
        /// Original declared event.
        event: EventRef,
        /// Declared dotted field path.
        path: String,
        /// Original field shape, including optionality.
        shape: PayloadShape,
        /// Explicit presence requirement when the original field is optional.
        require_present: bool,
    },
    /// Derive one exact signed integer from a preceding typed capture (suite/12+).
    Offset {
        /// Unique derived value binding.
        instance: InstanceName,
        /// Preceding integer binding.
        source: InstanceName,
        /// Checked signed addition.
        plus: i64,
    },
    /// Assert bounds on a previously selected occurrence, never as a filter.
    IntegerBounds {
        /// Previously observed occurrence.
        anchor: InstanceName,
        /// Original declared event.
        event: EventRef,
        /// Declared integer payload path.
        path: String,
        /// Original integer field shape.
        shape: PayloadShape,
        /// Inclusive lower bound.
        min: i64,
        /// Inclusive upper bound.
        max: i64,
    },
    /// Name the latest snapshot after a quiet interval fenced at activation.
    Quiet {
        /// Unique name for the selected actual occurrence.
        anchor: InstanceName,
        /// Full quiet interval after activation and the latest scoped receipt.
        duration_ms: u64,
        /// Account or resource scope, independent of the later captured count.
        claim: Claim,
    },
    /// Subscribe before any setup or stimulus.
    Open {
        /// Fixed complete set of source events for this scenario.
        events: BTreeSet<EventRef>,
    },
    /// Bind the first real matching occurrence after a preceding occurrence, if named.
    Await {
        /// Unique occurrence name, separate from response variables.
        anchor: InstanceName,
        /// Earlier occurrence, or subscription start.
        after: Option<InstanceName>,
        /// Payload assertion.
        claim: Claim,
    },
    /// No matching occurrence in the entire anchored interval.
    Absent {
        /// Earlier observed occurrence.
        anchor: InstanceName,
        /// Positive window length in milliseconds, exclusive at its end.
        duration_ms: u64,
        /// Forbidden event payload.
        claim: Claim,
    },
    /// Baseline and every scoped snapshot must agree throughout an interval.
    Stable {
        /// Earlier observed baseline.
        anchor: InstanceName,
        /// Positive window length in milliseconds.
        duration_ms: u64,
        /// Which account or resource snapshots belong to the claim.
        claim: Claim,
        /// Values checked on every scoped snapshot; cannot overlap the scope filter.
        required: BTreeMap<String, ScenarioValue>,
        /// Original types of asserted fields.
        shape: PayloadShape,
    },
    /// Require two distinct actual occurrences in strict order.
    Ordered {
        /// Earlier occurrence.
        before: InstanceName,
        /// Later occurrence.
        after: InstanceName,
    },
}

impl Check {
    /// Whether this check needs the captured-observation vocabulary.
    pub(crate) fn requires_v12(&self) -> bool {
        matches!(
            self,
            Self::Capture { .. }
                | Self::Offset { .. }
                | Self::IntegerBounds { .. }
                | Self::Quiet { .. }
        )
    }

    /// Events required from the target.
    pub fn events(&self) -> BTreeSet<EventRef> {
        match self {
            Self::Open { events } => events.clone(),
            Self::Await { claim, .. }
            | Self::Absent { claim, .. }
            | Self::Stable { claim, .. }
            | Self::Quiet { claim, .. } => [claim.event.clone()].into(),
            Self::Capture { event, .. } | Self::IntegerBounds { event, .. } => {
                [event.clone()].into()
            }
            Self::Ordered { .. } | Self::Offset { .. } => BTreeSet::new(),
        }
    }
}

/// Request for an append-only complete extension. Zero `through_ms` asks for new facts.
#[derive(Debug, Clone)]
pub struct Request {
    /// Exact subscribed lifetime.
    pub lifetime: String,
    /// Last accepted ordinal.
    pub after: u64,
    /// Minimum exclusive watermark required by a bounded assertion.
    pub through_ms: u64,
}

/// Native execution state, never accepted from the target or host report.
#[derive(Debug)]
pub struct State {
    lifetime: String,
    ledger: Ledger,
    anchors: BTreeMap<InstanceName, u64>,
    events: BTreeSet<EventRef>,
    bindings: BTreeMap<InstanceName, Node>,
    quiet: Option<QuietStart>,
}

#[derive(Debug)]
struct QuietStart {
    anchor: InstanceName,
    floor_ms: Option<u64>,
}
impl State {
    /// Begin a subscribed lifetime.
    pub fn new(lifetime: String, events: BTreeSet<EventRef>) -> Result<Self, temporal::Error> {
        Ok(Self {
            ledger: Ledger::new(lifetime.clone())?,
            lifetime,
            anchors: BTreeMap::new(),
            events,
            bindings: BTreeMap::new(),
            quiet: None,
        })
    }
    /// Native cursor, not an adapter-selected starting point.
    pub fn request(&self, through_ms: u64) -> Request {
        Request {
            lifetime: self.lifetime.clone(),
            after: self.ledger.cursor(),
            through_ms,
        }
    }
    /// Admit every fact in one batch; unexpected events are not silently dropped.
    pub fn accept(&mut self, batch: Batch) -> Result<(), temporal::Error> {
        if batch
            .occurrences
            .iter()
            .any(|item| !self.events.contains(&item.event))
        {
            return Err(self
                .ledger
                .invalidate("unsubscribed event in observation batch"));
        }
        self.ledger.accept(batch)?;
        if let Some(quiet) = self.quiet.as_mut() {
            if quiet.floor_ms.is_none() {
                quiet.floor_ms = Some(self.ledger.complete_before_ms());
            }
        }
        Ok(())
    }
    /// Publish the actual or derived value only after its native check succeeded.
    pub(crate) fn binding(&self, check: &Check) -> Option<(InstanceName, Node)> {
        match check {
            Check::Capture { instance, .. } | Check::Offset { instance, .. } => self
                .bindings
                .get(instance)
                .map(|value| (instance.clone(), value.clone())),
            _ => None,
        }
    }
    fn bind(&mut self, instance: &InstanceName, value: Node) -> Result<(), temporal::Error> {
        if self.bindings.insert(instance.clone(), value).is_some() {
            return Err(temporal::Error::Incomplete("duplicate live capture".into()));
        }
        Ok(())
    }
    fn anchor(&self, name: &InstanceName) -> Result<u64, temporal::Error> {
        self.anchors
            .get(name)
            .copied()
            .ok_or(temporal::Error::UnknownAnchor(0))
    }
    /// Evaluate a check from retained facts; Unfinished asks the evaluator for another batch.
    #[allow(
        clippy::too_many_lines,
        reason = "One exhaustive dispatch keeps all native receipt checks and their failures visible together."
    )]
    pub fn evaluate(
        &mut self,
        check: &Check,
        values: &BTreeMap<String, Node>,
        required: &BTreeMap<String, Node>,
    ) -> Result<(), temporal::Error> {
        match check {
            Check::Capture {
                anchor,
                instance,
                event,
                path,
                shape,
                ..
            } => {
                let item = self.ledger.anchor(self.anchor(anchor)?)?;
                let value = temporal::path_value(&item.payload, path);
                if &item.event != event
                    || value.is_none()
                    || value == Some(&Node::Null)
                    || shape
                        .leaves()
                        .get(path)
                        .is_none_or(|leaf| !leaf.admits(value))
                {
                    return Err(temporal::Error::Counterexample(item.clone()));
                }
                self.bind(instance, value.expect("checked actual field").clone())
            }
            Check::Offset { instance, plus, .. } => {
                let value = values
                    .get("base")
                    .and_then(|value| match value {
                        Node::Number(value) => value.as_i64(),
                        _ => None,
                    })
                    .ok_or_else(|| {
                        temporal::Error::Incomplete(
                            "offset requires a present signed integer".into(),
                        )
                    })?;
                let result = value.checked_add(*plus).ok_or_else(|| {
                    temporal::Error::Incomplete("signed integer offset overflow".into())
                })?;
                self.bind(instance, Node::Number(Number::from(result)))
            }
            Check::IntegerBounds {
                anchor,
                event,
                path,
                min,
                max,
                ..
            } => {
                let item = self.ledger.anchor(self.anchor(anchor)?)?;
                let value =
                    temporal::path_value(&item.payload, path).and_then(|value| match value {
                        Node::Number(value) => value.as_i64(),
                        _ => None,
                    });
                if &item.event != event
                    || !value.is_some_and(|value| *min <= value && value <= *max)
                {
                    return Err(temporal::Error::Counterexample(item.clone()));
                }
                Ok(())
            }
            Check::Quiet {
                anchor,
                duration_ms,
                claim,
            } => {
                if self.quiet.as_ref().map(|quiet| &quiet.anchor) != Some(anchor) {
                    self.quiet = Some(QuietStart {
                        anchor: anchor.clone(),
                        floor_ms: None,
                    });
                }
                let floor = self.quiet.as_ref().and_then(|quiet| quiet.floor_ms).ok_or(
                    temporal::Error::Unfinished {
                        required_ms: 0,
                        observed_ms: 0,
                    },
                )?;
                let item = self.ledger.quiet(
                    &Matcher {
                        event: claim.event.clone(),
                        payload: values.clone(),
                    },
                    floor,
                    *duration_ms,
                )?;
                if claim
                    .shape
                    .leaves()
                    .iter()
                    .any(|(path, leaf)| !leaf.admits(temporal::path_value(&item.payload, path)))
                {
                    return Err(temporal::Error::Counterexample(item.clone()));
                }
                self.anchors.insert(anchor.clone(), item.sequence);
                Ok(())
            }
            Check::Open { .. } => Err(temporal::Error::Incomplete("lifetime already open".into())),
            Check::Await {
                anchor,
                after,
                claim,
            } => {
                let after = after
                    .as_ref()
                    .map(|name| self.anchor(name))
                    .transpose()?
                    .unwrap_or(0);
                let matcher = Matcher {
                    event: claim.event.clone(),
                    payload: values.clone(),
                };
                let item =
                    self.ledger
                        .find(&matcher, after)?
                        .ok_or(temporal::Error::Unfinished {
                            required_ms: 0,
                            observed_ms: 0,
                        })?;
                if claim
                    .shape
                    .leaves()
                    .iter()
                    .any(|(path, leaf)| !leaf.admits(temporal::path_value(&item.payload, path)))
                {
                    return Err(temporal::Error::Counterexample(item.clone()));
                }
                self.anchors.insert(anchor.clone(), item.sequence);
                Ok(())
            }
            Check::Absent {
                anchor,
                duration_ms,
                claim,
            } => self.ledger.absent(
                &Matcher {
                    event: claim.event.clone(),
                    payload: values.clone(),
                },
                self.anchor(anchor)?,
                *duration_ms,
            ),
            Check::Stable {
                anchor,
                duration_ms,
                claim,
                ..
            } => self.ledger.stable(
                &Matcher {
                    event: claim.event.clone(),
                    payload: values.clone(),
                },
                required,
                self.anchor(anchor)?,
                *duration_ms,
            ),
            Check::Ordered { before, after } => self
                .ledger
                .ordered(self.anchor(before)?, self.anchor(after)?),
        }
    }
}

/// Reject malformed sequences before any target operation.
#[allow(
    clippy::too_many_lines,
    reason = "Validate the ordered trace vocabulary in one pass over anchors and typed bindings."
)]
pub(crate) fn admit(steps: &[ScenarioStep], major: u64) -> Result<(), String> {
    let mut opened = None;
    let mut anchors = BTreeMap::new();
    let mut captures = BTreeSet::new();
    let mut integer_captures = BTreeSet::new();
    for (index, step) in steps.iter().enumerate() {
        match step {
            ScenarioStep::CaptureResponse {
                instance,
                field,
                shape,
                ..
            } => {
                captures.insert(instance.clone());
                if shape.leaves().get(field).is_some_and(integer_leaf) {
                    integer_captures.insert(instance.clone());
                }
            }
            ScenarioStep::CaptureInstance { instance, .. }
            | ScenarioStep::EstablishEntity { instance, .. } => {
                captures.insert(instance.clone());
            }
            ScenarioStep::CheckLive { trace } => {
                if trace.requires_v12() && major < 12 {
                    return Err(
                        "captured observations and quiet baselines require suite/12 or /13".into(),
                    );
                }
                if major < 10 {
                    return Err("live occurrence checks require suite/10 or /11".into());
                }
                if let Check::Open { events } = trace {
                    if index != 0 || events.is_empty() {
                        return Err(
                            "one nonempty observation subscription must be the first step".into(),
                        );
                    }
                    opened = Some(events);
                    continue;
                }
                let events = opened.ok_or("live assertion before subscription")?;
                if !trace.events().is_subset(events) {
                    return Err("assertion event was not subscribed".into());
                }
                let known = |name: &InstanceName| {
                    if anchors.contains_key(name) {
                        Ok(())
                    } else {
                        Err("unknown or forward occurrence anchor".to_owned())
                    }
                };
                match trace {
                    Check::Capture {
                        anchor,
                        instance,
                        event,
                        path,
                        shape,
                        require_present,
                    } => {
                        if anchors.get(anchor) != Some(event) {
                            return Err(
                                "capture event differs from its actual occurrence anchor".into()
                            );
                        }
                        let leaf = single_leaf(shape, path)?;
                        if leaf.optional && !require_present {
                            return Err("optional event capture requires explicit presence".into());
                        }
                        if !captures.insert(instance.clone()) {
                            return Err("duplicate live value capture".into());
                        }
                        if integer_leaf(leaf) {
                            integer_captures.insert(instance.clone());
                        }
                    }
                    Check::Offset {
                        instance, source, ..
                    } => {
                        if !integer_captures.contains(source) {
                            return Err("offset requires a preceding typed integer capture".into());
                        }
                        if !captures.insert(instance.clone()) {
                            return Err("duplicate live value capture".into());
                        }
                        integer_captures.insert(instance.clone());
                    }
                    Check::IntegerBounds {
                        anchor,
                        event,
                        path,
                        shape,
                        min,
                        max,
                    } => {
                        if anchors.get(anchor) != Some(event)
                            || min > max
                            || !integer_leaf(single_leaf(shape, path)?)
                        {
                            return Err("integer bounds require their actual event, integer path and ordered range".into());
                        }
                    }
                    Check::Quiet {
                        anchor,
                        duration_ms,
                        claim,
                    } => {
                        if *duration_ms == 0 {
                            return Err("live quiet window must be positive".into());
                        }
                        admit_values(&claim.matches, &claim.shape, &captures)?;
                        if anchors
                            .insert(anchor.clone(), claim.event.clone())
                            .is_some()
                        {
                            return Err("duplicate occurrence anchor".into());
                        }
                    }
                    Check::Await {
                        anchor,
                        after,
                        claim,
                    } => {
                        if let Some(after) = after {
                            known(after)?;
                        }
                        admit_values(&claim.matches, &claim.shape, &captures)?;
                        if anchors
                            .insert(anchor.clone(), claim.event.clone())
                            .is_some()
                        {
                            return Err("duplicate occurrence anchor".into());
                        }
                    }
                    Check::Absent {
                        anchor,
                        duration_ms,
                        claim,
                    }
                    | Check::Stable {
                        anchor,
                        duration_ms,
                        claim,
                        ..
                    } => {
                        known(anchor)?;
                        if *duration_ms == 0 {
                            return Err("live window must be positive".into());
                        }
                        admit_values(&claim.matches, &claim.shape, &captures)?;
                        if let Check::Stable {
                            required, shape, ..
                        } = trace
                        {
                            admit_values(required, shape, &captures)?;
                            for a in required.keys() {
                                for b in claim.matches.keys() {
                                    if a == b
                                        || a.starts_with(&format!("{b}."))
                                        || b.starts_with(&format!("{a}."))
                                    {
                                        return Err(
                                            "stability claim overlaps its scope filter".into()
                                        );
                                    }
                                }
                            }
                        }
                    }
                    Check::Ordered { before, after } => {
                        known(before)?;
                        known(after)?;
                    }
                    Check::Open { .. } => unreachable!(),
                }
            }
            _ => {}
        }
    }
    Ok(())
}
pub(crate) fn integer_leaf(leaf: &crate::scenario::LeafShape) -> bool {
    matches!(
        leaf.holds,
        crate::scenario::Holds::Primitive {
            kind: ess_domain::types::Primitive::Integer
        }
    )
}
fn single_leaf<'a>(
    shape: &'a PayloadShape,
    path: &str,
) -> Result<&'a crate::scenario::LeafShape, String> {
    if shape.leaves().len() != 1 || path.is_empty() || path.split('.').any(str::is_empty) {
        return Err("event capture needs exactly one declared payload path".into());
    }
    shape
        .leaves()
        .get(path)
        .ok_or_else(|| "event capture has no declared field shape".into())
}
fn admit_values(
    values: &BTreeMap<String, ScenarioValue>,
    shape: &PayloadShape,
    captures: &BTreeSet<InstanceName>,
) -> Result<(), String> {
    if values.is_empty() {
        return Err("live claim must constrain a declared payload path".into());
    }
    for (path, value) in values {
        let leaf = shape
            .leaves()
            .get(path)
            .ok_or("live claim path has no declared type")?;
        match value {
            ScenarioValue::Literal { value } if leaf.admits(Some(value)) => {}
            ScenarioValue::Instance { instance } if captures.contains(instance) => {}
            _ => {
                return Err(
                    "live claim requires a typed literal or an earlier response capture".into(),
                );
            }
        }
    }
    Ok(())
}
